README.md
Cocorico Listing Search Bundle
==============================

# Installation

## Edit `composer.json`:
     
    ...
    "repositories": [
        {
          "type": "composer",
          "url": "https://packages.cocorico.io",
          "options": {
            "ssl": {
              "verify_peer": true,
              "allow_self_signed": true
            }
          }
        }
    ],
    ...

## Copy / paste auth.json.dist to auth.json and add Cocorico account in auth.json in "http-basic" part.

Use 0.1 version for Cocorico 0.1 branch and 0.2 for Cocorico 0.2 branch

    php composer.phar require cocorico/listing-search-bundle "^0.2"
    
    
## Edit `app/config/AppKernel.php` file:

    public function registerBundles()
    {
        $bundles = array(
            // ...
            new Cocorico\MangoPayBundle\CocoricoListingSearchBundle(),
            // ...
        );

        return $bundles;
    }

## Update `Cocorico/CoreBundle/Entity/Listing.php` file:

    ...
    @ORM\Table(name="listing",indexes={
    ...
    *   @ORM\Index(name="platform_notation_idx", columns={"platform_notation"}),
    ...
     
    class Listing extends BaseListing
    {
        ...
        use \Cocorico\ListingSearchBundle\Model\ListingSearchableTrait;
        ...
        

        
## Update schema:

Dry run:
    
   `php app/console doctrine:schema:update --dump-sql`
        
Update: 
    
   `php app/console doctrine:schema:update --force --env=dev`
        
        
## Add cron:

Listings platform notation computing:
        
    `30 2 * * * php <path-to-your-app>app/console cocorico_listing_search:computeNotation --env=dev`
    
        
## Optionally Edit `app/config/parameters.yml` file to adjust notation weight:
    ...
    cocorico_listing_search.notation_weights:
       listing_admin_notation: 5
       listing_fill_rate: 4
       listing_avg_rating: 5
       listing_availabilities_updated: 4
       listing_last_bookings: 4
       listing_certified: 4
       listing_new_bonus: 3
       listing_random_bonus: 2
       offerer_fill_rate: 3
       offerer_answers_rate: 4
       offerer_acceptation_rate: 4
       offerer_nb_bank_wires: 4
       offerer_answers_delay: 3
    ...        
            

CHANGELOG.md
CHANGELOG
=========

This changelog references the relevant changes done in this project.

This project adheres to [Semantic Versioning](http://semver.org/) 
and to the [CHANGELOG recommendations](http://keepachangelog.com/).


## [Unreleased]


## [0.2.2] - 2016-11-16

### Add

- Add search by distance
- Extend search when insufficient results


## [0.2.1] - 2016-05-31

### Changed

- Update doc and composer.json
- Remove stopPropagation from listingSearchSubscriber event


## [0.2.0] - 2016-04-01

### Added

- Support for Symfony 2.8
- Copyright


## [0.1.0] - 2016-03-17

### Added

- First commit: Bundle decoupled



Twig/ListingSearchExtension.php
<?php

/*
 * This file is part of the Cocorico package.
 *
 * (c) Cocolabs SAS <contact@cocolabs.io>
 *
 * For the full copyright and license information, please view the LICENSE
 * file that was distributed with this source code.
 */


namespace Cocorico\ListingSearchBundle\Twig;


class ListingSearchExtension extends \Twig_Extension implements \Twig_Extension_GlobalsInterface
{

    /**
     * @param string $listingSearchLocationType
     */
    public function __construct($listingSearchLocationType)
    {
        $this->listingSearchLocationType = $listingSearchLocationType;
    }


    public function getGlobals()
    {
        return array(
            'listingSearchLocationType' => $this->listingSearchLocationType,
        );
    }

    public function getName()
    {
        return 'cocorico_listing_search_extension';
    }
}


Model/ListingSearchableTrait.php
<?php

/*
 * This file is part of the Cocorico package.
 *
 * (c) Cocolabs SAS <contact@cocolabs.io>
 *
 * For the full copyright and license information, please view the LICENSE
 * file that was distributed with this source code.
 */

namespace Cocorico\ListingSearchBundle\Model;

/**
 * ListingSearchable trait.
 *
 * Should be used inside Listing entity, that needs to be searchable through an advanced algorithm.
 */
trait ListingSearchableTrait
{
    /**
     * Auto computed notation
     *
     * @ORM\Column(name="platform_notation", type="decimal", precision=15, scale=14, nullable=true)
     *
     * @var float
     */
    protected $platformNotation = 0;


    /**
     * @return float
     */
    public function getPlatformNotation()
    {
        return $this->platformNotation;
    }

    /**
     * @param float $platformNotation
     */
    public function setPlatformNotation($platformNotation)
    {
        $this->platformNotation = $platformNotation;
    }
}


Event/ListingSearchSubscriber.php
<?php

/*
 * This file is part of the Cocorico package.
 *
 * (c) Cocolabs SAS <contact@cocolabs.io>
 *
 * For the full copyright and license information, please view the LICENSE
 * file that was distributed with this source code.
 */

namespace Cocorico\ListingSearchBundle\Event;

use Cocorico\CoreBundle\Document\ListingAvailability;
use Cocorico\CoreBundle\Event\ListingSearchEvent;
use Cocorico\CoreBundle\Event\ListingSearchEvents;
use Cocorico\CoreBundle\Model\ListingSearchRequest;
use Cocorico\CoreBundle\Model\Manager\ListingSearchManager;
use Doctrine\ORM\QueryBuilder;
use Symfony\Component\EventDispatcher\EventSubscriberInterface;

class ListingSearchSubscriber implements EventSubscriberInterface
{
    protected $listingSearchManager;
    protected $locationType;
    protected $distanceViewportDivisions;
    protected $maxDistance;
    protected $distanceViewport;
    protected $listingSearchMinResult;

    /**
     * @param ListingSearchManager $listingSearchManager
     * @param array                $parameters
     */
    public function __construct(ListingSearchManager $listingSearchManager, $parameters)
    {
        $this->listingSearchManager = $listingSearchManager;
        $parameters = $parameters["parameters"];
        $this->locationType = $parameters["cocorico_listing_search_location_type"];
        $this->distanceViewportDivisions = $parameters["cocorico_listing_search_distance_viewport_divisions"];
        $this->maxDistance = $parameters["cocorico_listing_search_max_distance"];
        $this->listingSearchMinResult = $parameters["cocorico_listing_search_min_result"];
    }

    /**
     * Modify search query
     *
     * @param ListingSearchEvent $event
     * @return bool
     * @throws \Exception
     */
    public function onListingSearch(ListingSearchEvent $event)
    {
        $listingSearchRequest = $event->getListingSearchRequest();
        $queryBuilder = $event->getQueryBuilder();

        $queryBuilder->addSelect('partial l.{id, platformNotation}');

        $queryBuilder = $this->getQueryBuilderLocationPart($queryBuilder, $listingSearchRequest);

        //Dates availabilities (from MongoDB)
        $dateRange = $listingSearchRequest->getDateRange();
        if ($dateRange && $dateRange->getStart() && $dateRange->getEnd()) {
            if ($this->listingSearchManager->getListingDefaultStatus() == ListingAvailability::STATUS_AVAILABLE) {
                //Get listings really available for searched dates
                $listingsAvailable = $this->listingSearchManager->getListingsAvailability(
                    $dateRange,
                    $listingSearchRequest->getTimeRange(),
                    $listingSearchRequest->getFlexibility(),
                    null,
                    array(ListingAvailability::STATUS_AVAILABLE)
                );

                //Listings really available
                if (count($listingsAvailable)) {
                    $queryBuilder
                        ->addSelect("CASE WHEN l.id IN (:listingsAvailable) THEN 1 ELSE 0 END as really_available")
                        ->setParameter('listingsAvailable', array_keys($listingsAvailable));
                }
            }
        }

        //Order
        switch ($listingSearchRequest->getSortBy()) {
            case 'price':
                $queryBuilder->orderBy("l.price", "ASC");
                $queryBuilder->addOrderBy("l.platformNotation", "DESC");
                break;
            case 'distance':
                $queryBuilder->orderBy("distance", "ASC");
                $queryBuilder->addOrderBy("l.platformNotation", "DESC");
                break;
            default://recommended
                if ($this->locationType == 'distance') {
                    $queryBuilder->orderBy("distance_range", "ASC");
                } else {
                    $queryBuilder->orderBy("accuracy", "DESC");
                }

                if (isset($listingsAvailable) && count($listingsAvailable)) {
                    $queryBuilder->addOrderBy("really_available", "DESC");
                }
                $queryBuilder->addOrderBy("l.platformNotation", "DESC");
                $queryBuilder->addOrderBy("distance", "ASC");
                break;
        }

        $event->setQueryBuilder($queryBuilder);
//        $event->stopPropagation();
    }

    /**
     * @param QueryBuilder         $queryBuilder
     * @param ListingSearchRequest $listingSearchRequest
     * @param int                  $extensionLevel
     *
     * @return QueryBuilder
     */
    private function getQueryBuilderLocationPart(
        QueryBuilder $queryBuilder,
        ListingSearchRequest $listingSearchRequest,
        $extensionLevel = 1
    ) {
        $searchLocation = $listingSearchRequest->getLocation();

        if ($this->locationType == 'distance') {//search by distance
            $queryBuilderTmp = clone $queryBuilder;

            $queryBuilderTmp = $this->removeWhereClausePart($queryBuilderTmp);

            //Increase search distance if insuficiant result. Min viewport distance is 1km
            $distanceViewport = max($searchLocation->getViewportDiagonalDistance(), 1) * pow(2, $extensionLevel - 1);

            //New location relative where clause is done by distance
            $queryBuilderTmp
                ->andHaving('distance <= :viewportDistance')
                ->setParameter('viewportDistance', $distanceViewport);

            //Split viewport in X parts in which listings are groupe by distance
            $nbDivisions = 1;
            foreach ($this->distanceViewportDivisions as $distanceViewportParam => $nbDivisionsParam) {
                if ($distanceViewport <= $distanceViewportParam) {
                    $nbDivisions = $nbDivisionsParam;
                    break;
                }
            }

            //distance viewport is divided in $nbDivisions
            $distanceRange = round($distanceViewport / $nbDivisions, 4);
            $nbDivisions--;//The last case is greater than all others cases
            if (!$nbDivisions) {//nb division is equal to 0 if distance viewport is too small
                $distanceSQL = '1 AS distance_range';
            } else {
                $distanceSQL = '';
                for ($i = 0; $i < $nbDivisions; $i++) {
                    //todo: don't repeat GEO_DISTANCE here
                    $distanceSQL .= (!$i ? "CASE " : "") . "WHEN GEO_DISTANCE(co.lat = :lat, co.lng = :lng) <= :viewportDistance$i THEN $i " .
                        ($i == ($nbDivisions - 1) ? "ELSE $nbDivisions END AS distance_range " : "");
                    $queryBuilderTmp
                        ->setParameter("viewportDistance$i", ($i + 1) * $distanceRange);
                }
            }

            $queryBuilderTmp
                ->addSelect($distanceSQL);

            //Listings locations: The search location distance is extended while number of results is insuficiant
            if (count($queryBuilderTmp->getQuery()->getArrayResult()) < $this->listingSearchMinResult &&
                $distanceViewport < $this->maxDistance
            ) {
                $queryBuilderTmp = $this->getQueryBuilderLocationPart(
                    $queryBuilder,
                    $listingSearchRequest,
                    $extensionLevel + 1
                );
            }
            $queryBuilder = $queryBuilderTmp;
        } else {//search by viewport and location accuracy
            //Translated location ordering :
            //Listings having a location the most accurate compared to the searched one will be shown at first
            $accuracyMethod = $searchLocation->getAccuracyMethod();
            $accuracyMapping = $searchLocation->getAccuracyMapping();
            if ($accuracyMethod) {
                $accuracy = $searchLocation->$accuracyMethod();
                if ($accuracyMapping['table']) {
                    $queryBuilder
                        ->addSelect(
                            "CASE WHEN LOWER(amt." . $accuracyMapping['field'] . ") = :accuracy THEN 1 ELSE 0 END as accuracy"
                        )
                        ->leftJoin('co.' . $accuracyMapping['table'], 'am')
                        ->leftJoin('am.translations', 'amt')
                        ->andWhere('amt.locale = :locale');
                } else {
                    $queryBuilder->addSelect(
                        "CASE WHEN co." . $accuracyMapping['field'] . " = :accuracy THEN 1 ELSE 0 END as accuracy"
                    );
                }
                $queryBuilder->setParameter('accuracy', strtolower($accuracy));
            }
        }

        return $queryBuilder;
    }

    /**
     * Remove some part of queryBuilder where clause
     *
     * @param QueryBuilder $queryBuilder
     *
     * @return QueryBuilder
     */
    private function removeWhereClausePart(QueryBuilder $queryBuilder)
    {
        //See ssrc/Cocorico/CoreBundle/Model/Manager/ListingSearchManager.php:91
        $whereClausesToRemove = array('co.lat < :neLat', 'co.lat > :swLat', 'co.lng < :neLng', 'co.lng > :swLng');
        $whereParamsToRemove = array('neLat', 'swLat', 'neLng', 'swLng');

        //Remove location relative where clauses
        $whereParts = $queryBuilder->getDqlPart('where')->getParts();
        $queryBuilder->resetDQLPart('where');
        foreach ($whereParts as $wherePart) {
            if (in_array($wherePart, $whereClausesToRemove)) {
                continue;
            }
            $queryBuilder->andWhere($wherePart);
        }

        //Remove location relative where clauses parameters
        $parameters = $queryBuilder->getParameters();
        foreach ($parameters as $key => $parameter) {
            if (in_array($parameter->getName(), $whereParamsToRemove)) {
                $parameters->remove($key);
            }
        }
        $queryBuilder->setParameters($parameters);

        return $queryBuilder;
    }

    public static function getSubscribedEvents()
    {
        return array(
            ListingSearchEvents::LISTING_SEARCH => array('onListingSearch', 1),
        );
    }

}

DependencyInjection/Configuration.php
<?php

/*
 * This file is part of the Cocorico package.
 *
 * (c) Cocolabs SAS <contact@cocolabs.io>
 *
 * For the full copyright and license information, please view the LICENSE
 * file that was distributed with this source code.
 */

namespace Cocorico\ListingSearchBundle\DependencyInjection;

use Symfony\Component\Config\Definition\Builder\TreeBuilder;
use Symfony\Component\Config\Definition\ConfigurationInterface;

/**
 * This is the class that validates and merges configuration from your app/config files
 *
 * To learn more see {@link http://symfony.com/doc/current/cookbook/bundles/extension.html#cookbook-bundles-extension-config-class}
 */
class Configuration implements ConfigurationInterface
{
    /**
     * {@inheritdoc}
     */
    public function getConfigTreeBuilder()
    {
        $treeBuilder = new TreeBuilder();
        $rootNode = $treeBuilder->root('cocorico_listing_search');

        // Here you should define the parameters that are allowed to
        // configure your bundle. See the documentation linked above for
        // more information on that topic.

        return $treeBuilder;
    }
}


Command/ComputeListingsNotationCommand.php
<?php

/*
 * This file is part of the Cocorico package.
 *
 * (c) Cocolabs SAS <contact@cocolabs.io>
 *
 * For the full copyright and license information, please view the LICENSE
 * file that was distributed with this source code.
 */

namespace Cocorico\ListingSearchBundle\Command;

use Symfony\Bundle\FrameworkBundle\Command\ContainerAwareCommand;
use Symfony\Component\Console\Input\InputInterface;
use Symfony\Component\Console\Output\OutputInterface;

/*
 * Listings platform notation computing command
 * For example every day :
 */

//Cron: 30 2  * * *  user   php app/console cocorico_listing_search:computeNotation

class ComputeListingsNotationCommand extends ContainerAwareCommand
{

    public function configure()
    {
        $this
            ->setName('cocorico_listing_search:computeNotation')
            ->setDescription('Compute Listings Notation.')
            ->setHelp("Usage php app/console cocorico_listing_search:computeNotation");
    }

    public function execute(InputInterface $input, OutputInterface $output)
    {
        $container = $this->getContainer();
        $listingSearchManager = $container->get('cocorico_listing_search.manager');

        $result = $listingSearchManager->computeListingsNotation(
            $this->getContainer()->getParameter('cocorico.locale'),
            $this->getContainer()->getParameter("cocorico.listing_img_min"),
            $this->getContainer()->getParameter("cocorico.user_img_min")
        );

        $output->writeln($result . " listing(s) notation computed");
    }

}


DependencyInjection/CocoricoListingSearchExtension.php
<?php

/*
 * This file is part of the Cocorico package.
 *
 * (c) Cocolabs SAS <contact@cocolabs.io>
 *
 * For the full copyright and license information, please view the LICENSE
 * file that was distributed with this source code.
 */

namespace Cocorico\ListingSearchBundle\DependencyInjection;

use Symfony\Component\Config\FileLocator;
use Symfony\Component\DependencyInjection\ContainerBuilder;
use Symfony\Component\DependencyInjection\Loader;
use Symfony\Component\HttpKernel\DependencyInjection\Extension;

/**
 * This is the class that loads and manages your bundle configuration
 *
 * To learn more see {@link http://symfony.com/doc/current/cookbook/bundles/extension.html}
 */
class CocoricoListingSearchExtension extends Extension
{
    /**
     * {@inheritdoc}
     */
    public function load(array $configs, ContainerBuilder $container)
    {
        $configuration = new Configuration();
        $config = $this->processConfiguration($configuration, $configs);

        $loader = new Loader\YamlFileLoader($container, new FileLocator(__DIR__ . '/../Resources/config'));
        $loader->load('services.yml');
    }
}


Model/Manager/ListingSearchManager.php
<?php

/*
 * This file is part of the Cocorico package.
 *
 * (c) Cocolabs SAS <contact@cocolabs.io>
 *
 * For the full copyright and license information, please view the LICENSE
 * file that was distributed with this source code.
 */

namespace Cocorico\ListingSearchBundle\Model\Manager;


use Cocorico\CoreBundle\Entity\Booking;
use Cocorico\CoreBundle\Entity\Listing;
use Cocorico\CoreBundle\Repository\BookingRepository;
use Cocorico\CoreBundle\Repository\ListingRepository;
use Cocorico\MessageBundle\Model\ThreadManager;
use Doctrine\ORM\EntityManager;

class ListingSearchManager
{

    protected $em;
    protected $threadManager;
    protected $weights;

    /**
     * @param EntityManager $em
     * @param ThreadManager $threadManager
     * @param array         $weights
     */
    public function __construct(EntityManager $em, ThreadManager $threadManager, array $weights)
    {
        $this->em = $em;
        $this->threadManager = $threadManager;
        $this->weights = $weights;
    }

    /**
     * @param string $locale
     * @param int    $minListingImages
     * @param int    $minUserImages
     *
     * @return int
     */
    public function computeListingsNotation($locale, $minListingImages, $minUserImages)
    {
        /** @var ListingRepository $bookingRepository */
        $listingRepository = $this->em->getRepository('CocoricoCoreBundle:Listing');
        /** @var Listing[] $listings */
        $listings = $listingRepository->findBy(array("status" => Listing::STATUS_PUBLISHED));
        $globalTotalWeights =
            array_sum($this->weights) - $this->weights["listing_random_bonus"] - $this->weights["listing_new_bonus"];
        $nbVotesWeight = 3;//Significant nb votes
        $maxRating = 5;
        $nbListings = count($listings);
        //Random bonus
        $luckyListings = array_rand($listings, ceil(0.05 * $nbListings));
        if (!is_array($luckyListings)) {
            $luckyListings = array($luckyListings);
        }

        foreach ($listings as $i => $listing) {
            $totalWeights = $globalTotalWeights;
            $offerer = $listing->getUser();
            //----Admin notation
            $adminNotation = $listing->getAdminNotation() / 10;

            //----Listing fill rate
            $listingFillRate = array_sum($listing->getCompletionInformations($minListingImages, false)) / 5;

            //----User fill rate
            $userFillRate = array_sum($offerer->getCompletionInformations($minUserImages, false)) / 2;

            //----Rating: Rating value depends on number of votes
            $nbVotes = $listing->getCommentCount();
            $avgRating = $listing->getAverageRating();
            //More votes more the average rating is equal to the average rating
            $avgRating = (($avgRating * (1 - exp(-$maxRating * $nbVotes / $nbVotesWeight)) * 100) / 100) / 5;

            //----Calendar update
            $availabilitiesUpdatedAt = $listing->getAvailabilitiesUpdatedAt();
            $nbCalDays = 0;
            if ($availabilitiesUpdatedAt) {
                $now = new \DateTime();
                $nbCalDays = intval($availabilitiesUpdatedAt->diff($now)->format("%a"));
                $nbCalDays = $this->getNote("calDays", $nbCalDays);
            }

            //----Nb Bookings in last 30 days
            /** @var BookingRepository $bookingRepository */
            $bookingRepository = $this->em->getRepository('CocoricoCoreBundle:Booking');
            $now = new \DateTime();
            $lastBookings = $bookingRepository->findByListingAndLastCreated(
                $listing->getId(),
                $locale,
                array(Booking::STATUS_PAYED),
                $now->modify('-1 month')
            );
            $lastBookings = $this->getNote("lastBookings", count($lastBookings));

            //----Answers Ratings
            $answersRateAndDelay = $this->threadManager->getReplyRateAndDelay($offerer);
            $answersRate = $answersRateAndDelay["reply_rate"];

            //----Answers delay
            $answersDelay = $this->getNote("answersDelay", $answersRateAndDelay["reply_delay"]);

            $nbBookings = count($listing->getBookings());
            //----Acceptation Rate
            $bookingsAcceptedRate = 0;
            if ($nbBookings) {
                $bookingsAccepted = $bookingRepository->findByListingAndPayed(
                    $listing->getId(),
                    $locale
                );
                $bookingsAcceptedRate = count($bookingsAccepted) / $nbBookings;
            }

            //----bank Wires Rate
            $bankWiresRate = 0;
            if ($nbBookings) {
                $bookingsValidated = $bookingRepository->findByListingAndValidated(
                    $listing->getId(),
                    $locale
                );
                $bankWiresRate = (count($bookingsValidated) / $nbBookings);
            }

            //----Certified
            $certified = intval($listing->getCertified());

            //----Bonus new listing
            $listingCreatedAt = $listing->getCreatedAt();
            $now = new \DateTime();
            $listingNewBonus = 0;
            $newListingNbDays = intval($listingCreatedAt->diff($now)->format("%a"));
            if ($newListingNbDays <= 30) {
                $listingNewBonus = 1;
                $totalWeights += $this->weights["listing_new_bonus"];
            }

            //----Random Bonus
            $randomBonus = 0;
            if (in_array($i, $luckyListings)) {
                $randomBonus = mt_rand(1, 5) / 5;
                $totalWeights += $this->weights["listing_random_bonus"];
            }


            $notation = $adminNotation * $this->weights["listing_admin_notation"] +
                $listingFillRate * $this->weights["listing_fill_rate"] +
                $userFillRate * $this->weights["offerer_fill_rate"] +
                $avgRating * $this->weights["listing_avg_rating"] +
                $nbCalDays * $this->weights["listing_availabilities_updated"] +
                $lastBookings * $this->weights["listing_last_bookings"] +
                $answersRate * $this->weights["offerer_answers_rate"] +
                $bookingsAcceptedRate * $this->weights["offerer_acceptation_rate"] +
                $bankWiresRate * $this->weights["offerer_nb_bank_wires"] +
                $answersDelay * $this->weights["offerer_answers_delay"] +
                $certified * $this->weights["listing_certified"] +
                $listingNewBonus * $this->weights["listing_new_bonus"] +
                $randomBonus * $this->weights["listing_random_bonus"];


            $notation = $notation / $totalWeights;

            //Save notation
            $listing->setPlatformNotation($notation);

            // Save Answer delay
            $offerer->setAnswerDelay($answersRateAndDelay["reply_delay"]);

            // persist values
            $this->em->persist($listing);
            $this->em->persist($offerer);

//            echo("listing:" . $listing->getId() . PHP_EOL);
//            echo("globalTotalWeights:" . $globalTotalWeights . PHP_EOL);
//            echo("listing_admin_notation:" . $adminNotation . PHP_EOL);
//            echo("listing_fill_rate:" . $listingFillRate . PHP_EOL);
//            echo("offerer_fill_rate:" . $userFillRate . PHP_EOL);
//            echo("listing_avg_rating:" . $avgRating . PHP_EOL);
//            echo("listing_availabilities_updated:" . $nbCalDays . PHP_EOL);
//            echo("listing_last_bookings:" . $lastBookings . PHP_EOL);
//            echo("offerer_answers_rate:" . $answersRate . PHP_EOL);
//            echo("offerer_acceptation_rate:" . $bookingsAcceptedRate . PHP_EOL);
//            echo("offerer_nb_bank_wires:" . $bankWiresRate . PHP_EOL);
//            echo("offerer_answers_delay:" . $answersDelay . PHP_EOL);
//            echo("listing_certified:" . $certified . PHP_EOL);
//            echo("listing_new_bonus:" . $listingNewBonus . PHP_EOL);
//            echo("listing_random_bonus:" . $randomBonus . PHP_EOL);
//            echo("notation:" . $notation . PHP_EOL . PHP_EOL);

        }

        $this->em->flush();

        return $nbListings;
    }


    /**
     * Get note from value
     *
     * @param $type
     * @param $val
     * @return int
     */
    private function getNote($type, $val)
    {
        if ($type == "calDays") {
            if ($val <= 7) {
                $val = 5;
            } elseif ($val <= 15) {
                $val = 4;
            } elseif ($val <= 30) {
                $val = 3;
            } elseif ($val <= 45) {
                $val = 2;
            } elseif ($val <= 60) {
                $val = 1;
            } else {
                $val = 0;
            }
        } elseif ($type == "answersDelay") {
            if ($val > 0) {
                if ($val <= 7200) {//2h
                    $val = 5;
                } elseif ($val <= 14400) {//4h
                    $val = 4;
                } elseif ($val <= 43200) {//12h
                    $val = 3;
                } elseif ($val <= 86400) {//24h
                    $val = 2;
                } elseif ($val <= 129600) {//36h
                    $val = 1;
                } else {
                    $val = 0;
                }
            }
        } elseif ($type == "lastBookings") {
            if ($val >= 5) {
                $val = 5;
            } elseif ($val >= 4) {
                $val = 4;
            } elseif ($val >= 3) {
                $val = 3;
            } elseif ($val >= 2) {
                $val = 2;
            } elseif ($val >= 1) {
                $val = 1;
            } else {
                $val = 0;
            }
        } else {
            $val = 0;
        }

        return $val / 5;
    }
}